pkg/resolvers/k8s/watcher.go

```diff
@@ -61,112 +58,151 @@ func (w *watcher) Next() ([]*naming.Update, error) {
 	if err != nil {
 		// Just in case.
 		w.Close()
+		return u, errors.Wrap(err, "k8sresolver: ")
+	}
+	return u, nil
+}
+
+type key struct {
+	// EndpointAddress fields (usually pod).
+	ns, name string
+}
+
+func keyFromAddr(address v1.EndpointAddress) (key, error) {
+	if address.TargetRef == nil {
+		return key{}, errors.New("address targetRef is empty. Cannot maintain internal state.")
 	}
-	return u, err
+	return key{ns: address.TargetRef.Namespace, name: address.TargetRef.Name}, nil
 }
 
+// next gathers kube api endpoint watch changes and translates them to naming.Update set.
+// The main complexity is fact that naming.Update can be either Add or Delete. However kube events can be Add,
+// Delete or Modified. As a result we are required to maintain state. If the state is malformed we immdiately return error
+// which will cause resync on caller side (new resolver).
 func (w *watcher) next() ([]*naming.Update, error) {
-	updates := make([]*naming.Update, 0)
-	updatedEndpoints := make(map[string]struct{})
-	var event event
+	var (
+		updates           []*naming.Update
+		change            change
+		changeCh, errCh   = w.streamer.ResultChans()
+		endpointsToUpdate = map[key]string{}
+	)
+
 	select {
 	case <-w.ctx.Done():
 		// We already stopped.
 		return []*naming.Update(nil), w.ctx.Err()
-	case r := <-w.watchChange:
-		if r.err != nil {
-			return []*naming.Update(nil), errors.Wrap(r.err, "k8sresolver: error on reading event stream")
-		}
-		event = *r.ep
+	case err := <-errCh:
+		return []*naming.Update(nil), errors.Wrap(err, "error on reading change stream")
+	case change = <-changeCh:
 	}
 
-	// Translate kube api endpoint watch event to resolver address and put into map for easier lookup.
-	for _, subset := range event.Object.Subsets {
-		updatedAddresses, err := subsetToAddresses(w.target, subset)
+	for _, subset := range change.Subsets {
+		var err error
+		endpointsToUpdate, err = subsetToAddresses(w.target, subset)
 		if err != nil {
-			return []*naming.Update(nil), errors.Wrap(err, "k8sresolver: failed to convert k8s endpoint subset to update Addr")
+			return []*naming.Update(nil), errors.Wrap(err, "failed to convert k8s endpoint subset to update Addr")
 		}
 
-		for _, address := range updatedAddresses {
-			updatedEndpoints[address] = struct{}{}
+		if len(endpointsToUpdate) > 0 {
+			// Expected port found.
+			break
 		}
+
+		// Target port not found yet. Maybe other subsets includes target one?
 	}
 
-	// Create updates to add new endpoints.
-	for addr, md := range updatedEndpoints {
-		if _, ok := w.lastUpdates[addr]; ok {
-			continue
+	switch change.typ {
```

Comment:
Nit: you can put the loop outside to make it a bit simpler:
```
for k, addr := range endpointsToUpdate {
  oldAddr, hasOldAddr := w.endpoints[k]
  switch change.typ {...}
}
```

Reply:
you are right